base/accumulate.jl

```diff
@@ -442,3 +441,21 @@ function _accumulate1!(op, B, v1, A::AbstractVector, dim::Integer)
     end
     return B
 end
+
+# Internal function used to identify the widest possible eltype required for accumulate results
+function _accumulate_promote_op(op, v; init=nothing)
+    # Nested mock function that applies the `op` to each element of `v` which is necessary
+    # to find the widest return eltype.
+    # NOTE: We are just passing this to promote_op for inference and should never be run.
+    function f(op, v, init)
+        val, iter = Iterators.peel(v)
+        r = isnothing(init) ? reduce_first(op, val) : op(init, val)
```

Comment:
Yeah, the first option seems to give up and returns the wrong answer, but the second doesn't work because `k` needs to be defined... which is what the point of `r` is.

I really think the `promote_type(uniontypes(...)...)` has the most intuitive behaviour. It also handles the main edge case I was worried which was accidentally converting `Union{T, Missing/Nothing}` -> `Any`. Is there a particular performance issue or edge case you're concerned about?

Reply:
We don't actually run this code, so it doesn't matter if k actually gets defined

Reply:
Hmm, then I'm not sure why it always return `Any`... even in simple cases like `op=+` and `v=[1, 2, 3]` 😕

Reply:
Thanks, that does seem to work. So `k` does need to be defined even though the code isn't run? Is the preference for that solution simply the fact that it produces the narrowest possible answer, where `promote_union` will return `Any` when a `Union` would suffice?

Reply:
Yeah, my thought is that inference alone will generally pick a better type than `promote_union`, and that this seemed the most accurate model of what type we want it to compute

Reply:
yes, `k` needed to be defined in the right scope. I had defined it before too, but it had the wrong scope then

Reply:
Sorry, I was away from my workstation for a couple weeks. The two remaining issues with the approach you suggested is that:

1. It currently doesn't address the original `Union{}` case when `op(...)` would throw an error. This happens when `k` is undefined during the `convert(promote_type(...), r)` because the `op` errors in the loop. I'm fine with adding an early exit when `k` is undefined.

2. Inference seems to give up more easily in simple cases and causes other tests to fail.
```
Expression: cumsum(Any[1, 2.3]) == [1, 3.3] == cumsum(Real[1, 2.3])::Vector{Real}
  TypeError: in typeassert, expected Vector{Real}, got a value of type Vector{Any}
```
It's less clear to me how to fix this, or whether this is an acceptable compromise.

Reply:
yes, there could be an `isdefined` check there, but you have to be careful of how to write that because when written statically that check would always be true and result in skipping computing the desired promotion.

My other thought was that it might be possible to split `f` into 2 parts (before and after computing `k`) with the first one computing `typeof(k)`, then calling this `promote_type` directly (instead of inferring it)
```
        # next try to change `r` to that type
        r = convert(promote_type(typeof(r), typeof(k)), r)
```

My main concern with using `Base.promote_union` is that it can discard a lot of information and be quite terrible at computing the output if you have any unions in the input array. Whereas `promote_op` originally should be able to give a precise answer in those cases, which `promote_union` then discards.

Reply:
Hmm, it's not pretty, but it looks like breaking up `f` works.

1. Avoids the `isdefined` check
2. Fixes the test where inference was giving up and returning `Any`.

I'll push a commit, so you can take a look. Maybe there's a cleaner way to right it...?